Implement Referenced Token revocation on OAuth revocation endpoint

Ogenbertrand opened the pull request:
This PR delivers a comprehensive implementation of OAuth 2.0 compliant token revocation with full CORS support and seamless integration with the external status list server. The implementation transforms the Keycloak Token Status Plugin into a production-ready solution for distributed token lifecycle management. I created and compose and Dockerfile just for testing purposes, they will be removed after PR is closed.
 
 Closes: https://github.com/ADORSYS-GIS/token-status-link/issues/17

mbunwe-victor commented on src/main/java/com/adorsys/keycloakstatuslist/oauth/OAuthRevocationEndpoint.java:
```diff
@@ -0,0 +1,280 @@
+               "access_token".equals(tokenTypeHint) || 
+               "refresh_token".equals(tokenTypeHint);
+    }
+}
```
Could add a missing line here!

mbunwe-victor commented on src/main/java/com/adorsys/keycloakstatuslist/oauth/OAuthRevocationEndpointProvider.java:
```diff
@@ -0,0 +1,124 @@
+        
+        return record;
+    }
+}
```
Could add a missing line here!

mbunwe-victor commented on src/main/java/com/adorsys/keycloakstatuslist/oauth/OAuthRevocationResourceProvider.java:
```diff
@@ -0,0 +1,31 @@
+    public void close() {
+        // No resources to close
+    }
+}
```
Could add a missing line here!

mbunwe-victor commented on src/main/java/com/adorsys/keycloakstatuslist/oauth/OAuthRevocationResourceProviderFactory.java:
```diff
@@ -0,0 +1,46 @@
+    public String getId() {
+        return PROVIDER_ID;
+    }
+}
```
Could add a missing line here!

mbunwe-victor commented on src/test/java/com/adorsys/keycloakstatuslist/oauth/OAuthRevocationEndpointTest.java:
```diff
@@ -0,0 +1,172 @@
+        assertNotNull(response);
+        assertEquals(Response.Status.OK.getStatusCode(), response.getStatus());
+    }
+}
```
Could add a missing line here!

mbunwe-victor commented on src/test/java/com/adorsys/keycloakstatuslist/oauth/OAuthRevocationEndpointProviderTest.java:
```diff
@@ -0,0 +1,120 @@
+
+        assertNull(response); // Should return null to allow standard revocation to proceed
+    }
+}
```
Could add a missing line here!

mbunwe-victor commented on src/main/java/com/adorsys/keycloakstatuslist/oauth/OAuthRevocationEndpointProvider.java:
```diff
@@ -0,0 +1,124 @@
+    private static final Logger logger = Logger.getLogger(OAuthRevocationEndpointProvider.class);
+    
+    private final KeycloakSession session;
+    private final EventBuilder event;
```
Could you check the usage of the session and event defined, cause on my side it seems unused.

mbunwe-victor commented on src/main/java/com/adorsys/keycloakstatuslist/oauth/OAuthRevocationResourceProvider.java:
```diff
@@ -0,0 +1,31 @@
+public class OAuthRevocationResourceProvider implements RealmResourceProvider {
+
+    private final KeycloakSession session;
+    private final EventBuilder event;
```
The same thing here

Ogenbertrand replied:
They are used please check again.

Awambeng commented on src/test/java/com/adorsys/keycloakstatuslist/resource/CredentialRevocationResourceTest.java:
```diff
@@ -35,329 +36,271 @@ class CredentialRevocationResourceTest {
+    private CredentialRevocationService revocationService;
+    
+    @Mock
+    private org.keycloak.models.KeycloakSessionFactory sessionFactory;
```
Consider grouping related mocks together (e.g., session-related vs service-related) for readability.

Awambeng commented on src/test/java/com/adorsys/keycloakstatuslist/resource/CredentialRevocationResourceTest.java:
```diff
@@ -35,329 +36,271 @@ class CredentialRevocationResourceTest {
+            // Inject service
+            Field serviceField = CredentialRevocationResource.class.getDeclaredField("revocationService");
             serviceField.setAccessible(true);
             serviceField.set(resource, revocationService);
```
Reflection works but can be brittle. Consider constructor injection in the resource for easier testability and cleaner code.

You could extract header and service injection into helper methods to reduce boilerplate.

Awambeng commented on src/test/java/com/adorsys/keycloakstatuslist/resource/CredentialRevocationResourceTest.java:
```diff
@@ -35,329 +36,271 @@ class CredentialRevocationResourceTest {
+                throw new RuntimeException("Failed to access revocationService field", e);
+            }
+        }
     }
```
Reflection could throw runtime exceptions if fields are renamed. Constructor injection would remove this risk.

Awambeng commented on src/test/java/com/adorsys/keycloakstatuslist/resource/CredentialRevocationResourceTest.java:
```diff
@@ -35,329 +36,271 @@ class CredentialRevocationResourceTest {
-        verify(revocationService, never()).revokeCredential(any(), anyString());
+        assertNotNull(response);
+        assertEquals(200, response.getStatus());
+        verify(revocationService).revokeCredential(any(CredentialRevocationRequest.class), eq("test-token"));
```
You repeat similar mock setup in many tests. Consider a small helper method like mockRequest(token, reason, authHeader) to reduce duplication and improve readability.

Awambeng commented on src/main/java/com/adorsys/keycloakstatuslist/resource/CredentialRevocationResource.java:
```diff
@@ -1,176 +1,61 @@
-        }
-    }
+        if (authorizationHeader != null && authorizationHeader.toLowerCase().startsWith("bearer ")) {
+            String token = authorizationHeader.substring("bearer ".length()).trim();
```
Use equalsIgnoreCase("bearer") on split token type for robustness.

Consider logging when authorization header is missing or malformed for better observability.

Awambeng commented on src/main/java/com/adorsys/keycloakstatuslist/resource/CredentialRevocationResource.java:
```diff
@@ -1,176 +1,61 @@
+                    return super.revoke();
+
+                } catch (Exception e) {
+                    logger.errorf(e, "SD-JWT VP based revocation failed for credentialId: %s. Falling back to standard revocation.", credentialId);
```
Currently catching Exception broadly — you might want to catch specific exceptions if possible to avoid masking unexpected errors.

Awambeng reviewed (changes requested):
@Ogenbertrand I left a few comments. Could you please take a look?

Ogenbertrand commented on src/test/java/com/adorsys/keycloakstatuslist/resource/CredentialRevocationResourceTest.java:
```diff
@@ -1,363 +1,243 @@
+            String authorizationHeader = getHeaders().getHeaderString(HttpHeaders.AUTHORIZATION);
+
+            if (authorizationHeader != null && authorizationHeader.toLowerCase().startsWith("bearer ")) {
+                String token = authorizationHeader.substring("bearer ".length()).trim();
```
Your feedback regarding the authorization header parsing appears to have been addressed in the current version of the file, which already uses a robust method (.split() and equalsIgnoreCase).

IngridPuppet commented on src/main/java/com/adorsys/keycloakstatuslist/resource/CredentialRevocationResourceProviderFactory.java:
```diff
@@ -5,16 +5,10 @@
-    private static final Logger logger = Logger.getLogger(CredentialRevocationResourceProviderFactory.class);
-    private static final String PROVIDER_ID = "credential-revocation";
+    
+    public static final String PROVIDER_ID = "credential-revocation";
```
Please, what is the revocation endpoint according to this implementation, and what is the standard revocation endpoint?

IngridPuppet commented on src/main/java/com/adorsys/keycloakstatuslist/resource/CredentialRevocationResource.java:
```diff
@@ -1,176 +1,136 @@
+
+        if (authorizationHeader == null || authorizationHeader.trim().isEmpty()) {
+            logger.debug("No authorization header provided, falling back to standard revocation");
+            return super.revoke();
```
Do you mean standard revocation does not require an authorization header?

IngridPuppet reviewed (changes requested):
This implementation does not seem to fulfill the purpose of the ticket. Please could you check my comments?

Ogenbertrand commented:
I refactored credential revocation extension to eliminate custom endpoint (/credential-revocation) and integrate solely with Keycloak's standard /protocol/openid-connect/revoke endpoint via OID4VCLoginProtocolFactory override (ID: "openid-connect"). Removed CredentialRevocationResourceProvider; injected custom CredentialRevocationResource directly into factory's createTokenRevocationEndpoint. In revoke(): Trigger SD-JWT VP flow only on valid Bearer header + credentialId form param; return 200 OK directly on success; 400 "invalid_request" on missing credentialId; fallback to super.revoke() for no/malformed auth or custom errors (StatusListException/IllegalArgumentException). could you check this again @IngridPuppet ?

IngridPuppet commented on src/main/java/com/adorsys/keycloakstatuslist/resource/CredentialRevocationResourceProviderFactory.java:
```diff
@@ -26,37 +25,31 @@
     public String getId() {
-        return PROVIDER_ID;
+        return "openid-connect";
     }
```
No need to redefine this method just to return the same thing.

Ogenbertrand commented on src/main/java/com/adorsys/keycloakstatuslist/service/SdJwtVPValidationService.java:
I just pushed, could you check again @IngridPuppet !

Ogenbertrand commented on src/main/java/com/adorsys/keycloakstatuslist/service/SdJwtVPValidationService.java:
```diff
@@ -527,7 +524,6 @@ private KeyBindingJwtVerificationOpts getKeyBindingJwtVerificationOpts(String re
                     .withAllowedMaxAge(300)
-                    .withValidateExpirationClaim(true)   
                     .build();
         }
```
it's good to go this time :fire:

IngridPuppet reviewed (approved):
LGTM. Thx.
